src/DrCabal/Profile/Stacked.hs

```diff
@@ -104,8 +104,41 @@ entriesToPhase start times = Phase
             Nothing -> ct `minusw` start
             Just it -> ct `minusw` it
 
-formatChart :: Word64 -> Word64 -> Int -> Map Text Phase -> Text
-formatChart start end width libs = unlines $ concat
+{- | Take a log of Cabal events and derive a critical path, which is the longest
+stretch of dependent packages such that the sum of their build times determines
+the total wall clock time.
+
+The algorithm works this way:
+1. Find the very last event in the log and a corresponding library L.
+2. Find when L first emerges in the log (which is when its build started).
+3. If this is the very first entry of the log, exit.
+4. Find which build just finished before L's build started and a corresponding library L'.
+5. Set L := L' and return to Step 2.
+-}
+computeCriticalPath :: Map Text [(Status, Word64)] -> [Text]
+computeCriticalPath libs = case Map.maxView libsByFinishTime of
+    Nothing -> []  -- oh well, seems the log was empty
+    Just (lastLib, _) -> lastLib : unfoldr go lastLib
+  where
+    -- Inner lists are generated by 'groupEntries', so must be nonempty.
+    libsList :: [(Text, [Word64])]
+    libsList = map (fmap (map snd)) $ Map.assocs libs
+
+    libsByStartTime :: Map Text Word64
+    libsByStartTime = Map.fromList $ map (fmap List.minimum) libsList
+
+    libsByFinishTime :: Map Word64 Text
+    libsByFinishTime = Map.fromList $ map (swap . fmap List.maximum) libsList
```

Comment:
There's a minor inconsistency in naming. Both values are `libsBy...` but one is one `Map Text Word64` and another is `Map Word64 Text`. Let's rename `libsByFinishTime` to `finishTimeToLibs`

I imagine, this code relies on the assumption that no two entries should have the same time (which is probably a pragmatic assumption). But it could be something to document as well for future ourselves